### PR TITLE
Add Policy Groups deep dive

### PR DESCRIPTION
**Pull request by viccuad**
## Description

Fix https://github.com/kubewarden/kubewarden.io/issues/262

### CODE REVIEW

**Comment by viccuad** [comment]
Left in draft as we need to wait a bit after release. Publishing date to today so it is rendered.

**Review by fabriziosestito — APPROVED** [review]

File: content/blog/2024/10/policy-groups.md [hunk 1]
```diff
@@ -0,0 +1,276 @@
+---
+title: Policy Groups deep dive
+authors:
+  - Víctor Cuadrado Juan
+date: 2024-11-04
+---
+
+With [v1.17](https://www.kubewarden.io/blog/2024/10/kubewarden-1-17-release/),
+we introduced a new powerful feature, Policy Groups, brought by 2 new Kubernetes
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:9]:
```suggestion
we introduced a new powerful feature, Policy Groups, enabled by 2 new Kubernetes
```

File: content/blog/2024/10/policy-groups.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+- **AdmissionPolicyGroups**: Namespaced policy comprised of several policies.
+- **ClusterAdmissionPolicyGroups**: Clusterwide policy comprised of several policies.
+
+These new Policy Groups resources define a policy comprised of several policies and
+their policy settings, and they perform a combined evaluation of those multiple
+policies using logical operators.
+
+Why are these useful? Because they reuse existing policies, reducing the need
+for custom policy creation. And they provide complex logic while at the same
+time reducing complexity as you have all the logic packed in 1 resource
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:21]: I suggest 'contained' as 'packed in' does not feel like reducing complexity. As in lots of things packed in. Also we try to use written numbers up to ten. Not particularly important, but a stylistic convention.
```suggestion
time reducing complexity as you have all the logic contained in one resource
```

File: content/blog/2024/10/policy-groups.md [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+definition.
+
+## Writing a Policy Group
+
+AdmissionPolicyGroups and ClusterAdmissionPolicyGroups are similar to the
+policies you already knew. Let's start writing a ClusterAdmissionPolicyGroup
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:27]:
```suggestion
policies you already know. Let's start writing a ClusterAdmissionPolicyGroup
```

File: content/blog/2024/10/policy-groups.md [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+The expression evaluation has short-circuit and only evaluates those that are
+needed; if one of the boolean results is true and nothing else is needed to
+accept or reject, the other evaluations are not performed to save resources.
+
+The short-circuit means that we don't support mutation for Policy Groups so
+far, which simplifies evaluation and mental load.
+
+### Message field
+
+Ok, so we know how the policy gets evaluated into a rejection or acceptation.
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:104]:
```suggestion
Ok, so we know how the policy gets evaluated and rejected or accepted.
```

File: content/blog/2024/10/policy-groups.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+But how do we express that to the users? We do it by setting its
+`spec.message`, which gets returned as part of the AdmissionResponse as usual,
+to kubectl, or whomever expects it.
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:107]:
```suggestion
to kubectl, or whoever expects it.
```

File: content/blog/2024/10/policy-groups.md [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+```console
+$ kubectl apply -f unique-service-selector.yml
+clusteradmissionpolicygroup.policies.kubewarden.io/unique-service-selector created
+$ kubectl get clusteradmissionpolicygroups.policies.kubewarden.io
+NAME                      POLICY SERVER   MUTATING   BACKGROUNDAUDIT   MODE      OBSERVED MODE   STATUS    AGE
+unique-service-selector   default                    true              protect   protect         active    30s
+```
+
+On Policy Group instantiation, both the `spec.expression` and each of the policies'
+settings get validated, and if any is incorrect, one would get an error as usual.
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:165]:
```suggestion
settings get validated, and if any is incorrect, one gets an error as expected.
```

File: content/blog/2024/10/policy-groups.md [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+### Audit scanner
+
+Another way of obtaining details on the state of the cluster is the [Audit
+Scanner](https://docs.kubewarden.io/next/explanations/audit-scanner) feature.
+Policy Groups are fully supported in the Kubewarden stack, and their results
+get published to Policy Reports.
+
+## Recap
+
+We have seen the 2 new Policy Groups Custom Resources: **\*AdmissionPolicyGroups** and
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:259]: Did you want that asterisk in there?

File: content/blog/2024/10/policy-groups.md [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+  },
+  "code": 400
+}]
+Error from server: error when creating "second-service.yml": admission webhook "clusterwide-group-unique-service-selector.kubewarden.admission" denied the request: the service selector is not unique or the service is not owned by the Quokkas team
+```
+
+### Audit scanner
+
+Another way of obtaining details on the state of the cluster is the [Audit
+Scanner](https://docs.kubewarden.io/next/explanations/audit-scanner) feature.
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:253]:
```suggestion
Scanner](https://docs.kubewarden.io/next/explanations/audit-scanner).
```

File: content/blog/2024/10/policy-groups.md [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+  rules:
+    - apiGroups: [""]
+      apiVersions: ["v1"]
+      resources: ["services"]
+      operations: ["CREATE", "UPDATE"]
+```
+
+### Policies field & context-aware
+
+Now, we define several policies and their settings. This is done in its
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:48]:
```suggestion
Now, we define several policies and their settings. This is done in the
```

File: content/blog/2024/10/policy-groups.md [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+      mandatory_annotations:
+        - owner
+      constrained_annotations:
+        owner: "quokkas-team"
+```
+
+In this map, you can see we have defined 2 policies. One that checks for
+Services to be unique, and another that checks for an annotation `owner:
+foo-team`. These policy entries in the map are named. And this allows us
+to write a boolean expresion with the boolean results of each policy.
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:72]:
```suggestion
to write a boolean expression combining the results of each policy.
```

File: content/blog/2024/10/policy-groups.md [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+
+If you look even closer, you will realize that the `unique_service_selector`
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:74]:
```suggestion
If you look closer, you will see that the `unique_service_selector`
```

File: content/blog/2024/10/policy-groups.md [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+
+We evaluate that logic in the `spec.expression`. In our case:
+
+```yaml
+expression: "unique_service_selector() || (!unique_service_selector() && owned_by_quokkas_team())"
+```
+
+Which means that we expect the Service selector to be unique, or if it isn't,
+to be owned by the Quokkas team. The Quokkas team is awesome; they know how to
+have a clean house by themselves.
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:89]: Maybe?
```suggestion
keep a secure house.
```

File: content/blog/2024/10/policy-groups.md [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+
+The `expression` is a boolean one, evaluating the known policy identifiers
+defined in the Policy Group. We can use `&&`, `||`, `!` for AND, OR, and NOT
+operations, as well as `(`,`)` for evaluation priorities.
+
+The expression evaluation has short-circuit and only evaluates those that are
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:95]:
```suggestion
The expression evaluation has a short-circuit and only evaluates those that are
```

File: content/blog/2024/10/policy-groups.md [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+needed; if one of the boolean results is true and nothing else is needed to
+accept or reject, the other evaluations are not performed to save resources.
+
+The short-circuit means that we don't support mutation for Policy Groups so
+far, which simplifies evaluation and mental load.
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:100]:
```suggestion
The short-circuit means that we don't support mutation for Policy Groups, which simplifies evaluation and the conceptual load.
```

File: content/blog/2024/10/policy-groups.md [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+
+### Message field
+
+Ok, so we know how the policy gets evaluated into a rejection or acceptation.
+But how do we express that to the users? We do it by setting its
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:105]:
```suggestion
But how do we express that to users? We do it by setting its
```

File: content/blog/2024/10/policy-groups.md [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+`spec.message`, which gets returned as part of the AdmissionResponse as usual,
+to kubectl, or whomever expects it.
+
+In our case, we can write:
+
+```yaml
+message: "the service selector is not unique or the service is not owned by the foo team"
+```
+
+In addition, all the evaluation details of each the group policies are sent as part
```
Comment (jhkrug) [inline content/blog/2024/10/policy-groups.md:115]:
```suggestion
In addition, all the evaluation details of each of the group policies are sent as part
```

**Review by jhkrug — CHANGES_REQUESTED** [review]
Some suggestions for your consideration.

File: content/blog/2024/10/policy-groups.md (hunk 7 above)
Comment (viccuad) [reply content/blog/2024/10/policy-groups.md]: Nope, good catch! Fixed with https://github.com/kubewarden/kubewarden.io/pull/266/commits/85bfbba5feb64d8f20b4dcb8aba1f14e024e075a.

**Comment by flavio** [comment]
Moving to blocked, planning to publish that on Wed Oct 2nd

**Pull request merged**